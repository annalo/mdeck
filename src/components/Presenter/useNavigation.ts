import { useCallback, useEffect, useState } from "react";
import * as R from "ramda";

type SlideNumber = number;
interface UseNavigationProps {
  slideCount: number;
}

function useNavigation({ slideCount }: UseNavigationProps): SlideNumber {
  const [currentSlide, setCurrentSlide] = useState<SlideNumber>(1);

  const pageUp = useCallback(() => {
    setCurrentSlide((slideNumber) =>
      R.ifElse(R.equals(slideCount), R.identity, R.add(1))(slideNumber)
    );
  }, [setCurrentSlide, slideCount]);
  const pageDown = useCallback(() => {
    setCurrentSlide((slideNumber) =>
      R.ifElse(R.equals(1), R.identity, R.subtract(1))(slideNumber)
    );
  }, [setCurrentSlide]);

  useEffect(() => {
    const handleKeyDown = (e) => {
      switch (e.keyCode) {
        case 37:
        case 38: {
          pageDown();
          break;
        }
        case 39:
        case 40: {
          pageUp();
          break;
        }
        default: {
          break;
        }
      }
    };

    document.addEventListener("keydown", handleKeyDown);
    return () => {
      document.removeEventListener("keydown", handleKeyDown);
    };
  }, [currentSlide, pageUp, pageDown]);

  return currentSlide;
}

export { useNavigation };
