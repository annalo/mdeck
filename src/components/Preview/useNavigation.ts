import { useCallback, useEffect, useState } from "react";
import screenfull, { Screenfull } from "screenfull";
import scrollIntoView from "smooth-scroll-into-view-if-needed";

import { useSlideEntries } from "contexts/SlideObserver";

function useNavigation(): () => void {
  const [presentationMode, togglePresentation] = useState(false);
  const [slideNumber, setSlideNumber] = useState<SlideNumber | null>(null);

  const slideEntries = useSlideEntries();
  const slideCount = Object.keys(slideEntries).length;

  const requestPresentation = useCallback(() => {
    const slideshow = document.querySelector("#slideshow");
    if (slideshow)
      (screenfull as Screenfull).request(slideshow).then(() => {
        console.log("request fullscreen");
        togglePresentation(true);
      });
  }, []);

  useEffect(() => {
    const fullScreen = screenfull as Screenfull;
    const exitFullscreen = () => {
      if (!fullScreen.isFullscreen) togglePresentation(false);
    };
    fullScreen.on("change", exitFullscreen);
    return () => fullScreen.off("change", exitFullscreen);
  }, [togglePresentation]);

  useEffect(() => {
    console.log("use effect set slide number 1 when presentationMode");
    presentationMode ? setSlideNumber(1) : setSlideNumber(null);
  }, [presentationMode]);

  useEffect(() => {
    console.log("useeffect slidenumber changed", slideNumber);
    if (slideNumber) {
      const slide = slideEntries[slideNumber];
      scrollIntoView(slide);
    }
  }, [slideNumber, slideEntries]);

  const nextSlide = useCallback(
    () =>
      setSlideNumber((current) =>
        current && current < slideCount ? current + 1 : current
      ),
    [slideCount]
  );
  const previousSlide = useCallback(
    () =>
      setSlideNumber((current) =>
        current && current > 1 ? current - 1 : current
      ),
    []
  );

  useEffect(() => {
    const handleKeyDown = (e) => {
      console.log("handlekeydown");
      switch (e.keyCode) {
        case 37: // ArrowLeft
        case 38: // ArrowUp
          previousSlide();
          break;

        case 39: // ArrowRight
        case 40: // ArrowDown
          nextSlide();
          break;

        default: {
          break;
        }
      }
    };

    presentationMode
      ? document.addEventListener("keydown", handleKeyDown)
      : document.removeEventListener("keydown", handleKeyDown);

    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [nextSlide, previousSlide, presentationMode]);

  return requestPresentation;
}

export { useNavigation };
