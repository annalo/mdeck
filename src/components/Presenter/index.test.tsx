import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { act, renderHook } from "@testing-library/react-hooks";

import { Presenter } from ".";
import { useNavigation } from "./useNavigation";

describe("<Presenter />", () => {
  const htmlArray = [
    "<svg><h1>Title</h1></svg>",
    "<svg><p>Paragraph</p></svg>",
  ];
  const component = <Presenter htmlArray={htmlArray} />;

  test("should render and match the snapshot", () => {
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render one slide at a time ", () => {
    const { container } = render(component);
    expect(container).toContainHTML(htmlArray[0]);
    expect(container).not.toContainHTML(htmlArray[1]);
  });

  describe("useNavigation", () => {
    test("should not set currentSlide less than 1", async () => {
      const { result } = renderHook(() => useNavigation({ slideCount: 7 }));

      expect(result.current).toBe(1);
      act(() => {
        fireEvent.keyDown(document, { key: "ArrowUp", keyCode: 38 });
      });
      expect(result.current).toBe(1);
    });

    test("should not set currentSlide more than slideCount", async () => {
      const { result } = renderHook(() => useNavigation({ slideCount: 2 }));

      expect(result.current).toBe(1);
      act(() => {
        fireEvent.keyDown(document, { key: "ArrowDown", keyCode: 40 });
      });
      expect(result.current).toBe(2);
      act(() => {
        fireEvent.keyDown(document, { key: "ArrowDown", keyCode: 40 });
      });
      expect(result.current).toBe(2);
    });

    test("should navigate to next slide on ArrowDown", async () => {
      const { container } = render(component);

      expect(container).toContainHTML(htmlArray[0]);

      fireEvent.keyDown(document, { key: "ArrowDown", keyCode: 40 });

      await waitFor(() => {
        expect(container).toContainHTML(htmlArray[1]);
        expect(container).not.toContainHTML(htmlArray[0]);
      });
    });

    test("should navigate to next slide on ArrowRight", async () => {
      const { container } = render(component);

      expect(container).toContainHTML(htmlArray[0]);

      fireEvent.keyDown(document, { key: "ArrowRight", keyCode: 39 });

      await waitFor(() => {
        expect(container).toContainHTML(htmlArray[1]);
        expect(container).not.toContainHTML(htmlArray[0]);
      });
    });

    test("should navigate to previous slide on ArrowUp", async () => {
      const { container } = render(component);

      fireEvent.keyDown(document, { key: "ArrowDown", keyCode: 40 });
      await waitFor(() => expect(container).toContainHTML(htmlArray[1]));

      fireEvent.keyDown(document, { key: "ArrowUp", keyCode: 38 });
      await waitFor(() => {
        expect(container).toContainHTML(htmlArray[0]);
        expect(container).not.toContainHTML(htmlArray[1]);
      });
    });

    test("should navigate to previous slide on ArrowLeft", async () => {
      const { container } = render(component);

      fireEvent.keyDown(document, { key: "ArrowDown", keyCode: 40 });
      await waitFor(() => expect(container).toContainHTML(htmlArray[1]));

      fireEvent.keyDown(document, { key: "ArrowLeft", keyCode: 37 });
      await waitFor(() => {
        expect(container).toContainHTML(htmlArray[0]);
        expect(container).not.toContainHTML(htmlArray[1]);
      });
    });
  });
});
