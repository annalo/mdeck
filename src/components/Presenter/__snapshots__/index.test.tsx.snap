// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Presenter /> should render and match the snapshot 1`] = `
<DocumentFragment>
  <article
    id="presenter"
  >
    <div
      id="slide-1"
    >
      <svg>
        <h1>
          Title
        </h1>
      </svg>
    </div>
  </article>
</DocumentFragment>
`;
